proposed/0000-UDP-hole-punching.md

```diff
@@ -0,0 +1,192 @@
+                            peer_addr : mut SocketAddr /* of node B */,
+                            callback: FnBox<(UdpSocket, io::Result<SocketAddr>)>)
+
+This call will initiate reading on the `udp_socket` and will also
+start periodically sending small datagrams to the `peer_addr`.
+These datagrams will be of type:
+
+    struct HolePunch {
+        secret: Option<[u8; 4]>,
+        ack: bool,
+    }
+
+At this stage, we may receive the above datagram either from `peer_addr`
+or from some other endpoint. If the later, it can mean one of two things:
+
+* The other end is behind a Symmetric NAT and we're behind a Full-cone or no NAT.
+  This is OK but we need to adjust the `peer_addr` variable to point
+  to this new peer.
+* Some other application is sending us irrelevant or malicious packets. We
+  can distinguish this case by matchin the `secret` passed to the function as an
+  argument with `datagram.secret`. If secrets don't match, we ignore the datagram.
+
+Once we receive a datagram from the remote peer, we know our hole has been
+punched. Since the other end is waiting for our datagram as well, we need
+to make a "best effort" to ensure the other end receives our packet as well.
+That is, we need to make sure at least one of the following conditions is true:
+
+1. We sent our datagram to `peer_addr` `K` times (for some predefined `K`).
+2. We receive `HolePunch` packet with the `ack` field set to true.
+
+Note that the condition ((1) or (2)) will always be satisfied eventually,
+so the decision whether hole punching was successfull or not shall
+be based purely on the fact whether we have received a datagram or not, that is,
+once ((1) or (2)) is true, we check whether we've received a datagram,
+if we did, we call the `callback` with `(udp_socket, Ok(peer_addr))`,
+otherwise we call it with an appropriate `io::Error`.
+
+## Using the punched hole
+
+After a successful hole-punch, it is assumed that the two peers can communicate
+with each other over the unreliable UDP protocol. At this point we'd like to
+use that UDP socket with protocols such as uTP for added reliability.
+
+Unfortunatelly, the rust-utp library doesn't currently support rendezvous
+connections, so this functionality will need to be added.
+
+# Drawbacks
+
+The alternative approach described in the next section is simpler in that it doesn't require
+two async calls, instead, it uses only one. Other than that, the simpler approach
+is limited in number of NAT types it can successfully punch through.
+
+# Alternatives
+
+Another approach would work with the use of multiplexing. That is, if we had a UDP socket
+which is already connected to a remote peer, we could also use it to communicate
+with other peers as the hole has already been punched. Problem with this approach is 
+that it would only help with `Full-cone NAT` types because it is the only NAT
+type where a hole punched to one peer/host can be reused with another peers/hosts.
+ resulting in slower code (re
```

Comment:
Resolved in https://github.com/inetic/rfcs/commit/7db7a35d06cea9bbe6875a4bd844c05a57f78a70